File: code-python/net/socket_server_v5.py
import socket
import struct
import threading

def encode_tlv(type_, value):
    length = len(value)
    return struct.pack("!BH", type_, length) + value.encode()

def decode_tlv(buffer):
    if len(buffer) < 3:
        return None, buffer
    type_, length = struct.unpack("!BH", buffer[:3])
    if len(buffer) < 3 + length:
        return None, buffer
    value = buffer[3:3+length].decode()
    return (type_, value), buffer[3+length:]

def handle_client(conn, addr):
    print(f"Client connected: {addr}")
    conn.settimeout(30)
    
    buffer = b""
    
    while True:
        try:
            data = conn.recv(1024)
            if not data:
                break
            buffer += data
            while True:
                packet, buffer = decode_tlv(buffer)
                if not packet:
                    break
                type_, value = packet
                print(f"Received: Type={type_}, Value={value}")
                conn.settimeout(30)
        except socket.timeout:
            print("Connection timeout")
            break

    conn.close()
    print("Client disconnected")

server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
server.bind(("0.0.0.0", 8080))
server.listen(5)
print("Server listening on port 8080...")

while True:
    conn, addr = server.accept()
    threading.Thread(target=handle_client, args=(conn, addr)).start()

File: code_javascript/net/socket_server_v5.js
const net = require("net");

function encodeTLV(type, value) {
    let buffer = Buffer.alloc(3 + value.length);
    buffer.writeUInt8(type, 0);
    buffer.writeUInt16BE(value.length, 1);
    buffer.write(value, 3);
    return buffer;
}

function decodeTLV(buffer) {
    if (buffer.length < 3) return null;
    let type = buffer.readUInt8(0);
    let length = buffer.readUInt16BE(1);
    if (buffer.length < 3 + length) return null;
    return { type, value: buffer.toString("utf8", 3, 3 + length), remaining: buffer.slice(3 + length) };
}

const server = net.createServer((socket) => {
    console.log("Client connected");
    socket.setTimeout(30000);

    const interval = setInterval(() => {
        if (socket.destroyed) {
            clearInterval(interval);
        } else {
            socket.write("PING\n");
        }
    }, 10000); // 每 10 秒发送心跳

    let buffer = Buffer.alloc(0);

    socket.on("data", (data) => {
        buffer = Buffer.concat([buffer, data]);
        while (true) {
            let packet = decodeTLV(buffer);
            if (!packet) break;
            console.log(`Received: Type=${packet.type}, Value=${packet.value}`);
            buffer = packet.remaining;
        }
    });

    socket.on("timeout", () => {
        console.log("Connection timeout");
        socket.end();
    });

    
     
    socket.on("end", () => {
        clearInterval(interval);
        console.log("Client disconnected");
    });
});

server.listen(8080, () => console.log("Server listening on port 8080..."));
